yarn-project/noir-contracts/src/contracts/token_contract/src/main.nr

```diff
@@ -0,0 +1,364 @@
+        let supply = SafeU120::new(storage.total_supply.read()).add(amount);
+
+        storage.public_balances.at(to.address).write(new_balance.value as Field);
+        storage.total_supply.write(supply.value as Field);
+        1
+    }
+
+    #[aztec(public)]
+    fn mint_priv(
+        amount: Field,
+        secret_hash: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+        assert(storage.minters.at(context.msg_sender()).read() == 1, "caller is not minter");
+        let pending_shields = storage.pending_shields;
+        let mut note = TransparentNote::new(amount, secret_hash);
+        let supply = SafeU120::new(storage.total_supply.read()).add(SafeU120::new(amount));
+
+        storage.total_supply.write(supply.value as Field);
+        pending_shields.insert_from_public(&mut note);
+        1
+    }
+
+    #[aztec(public)]
+    fn shield(
+        from: AztecAddress,
+        amount: Field,
+        secret_hash: Field,
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+
+        if (from.address != context.msg_sender()) {
+            // The redeem is only spendable once, so we need to ensure that you cannot insert multiple shields from the same message.
+            let selector = compute_selector("shield((Field),Field,Field,Field)");
+            let message_field = compute_message_hash([context.msg_sender(), selector, from.address, amount, secret_hash, nonce]);
+            assert(AccountContract::at(from.address).is_valid(Option::none(), Option::some(context), message_field) == 0xe86ab4ff, "invalid call");
+            context.push_new_nullifier(message_field, 0);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let amount = SafeU120::new(amount);
+        let from_balance = SafeU120::new(storage.public_balances.at(from.address).read()).sub(amount);
+
+        let pending_shields = storage.pending_shields;
+        let mut note = TransparentNote::new(amount.value as Field, secret_hash);
+
+        storage.public_balances.at(from.address).write(from_balance.value as Field);
+        pending_shields.insert_from_public(&mut note);
+        1
+    }
+
+    #[aztec(public)]
+    fn transfer_public(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+    ) -> Field {
+        let storage = Storage::init(Option::none(), Option::some(&mut context));
+
+        if (from.address != context.msg_sender()) {
+            let selector = compute_selector("transfer_public((Field),(Field),Field,Field)");
+            let message_field = compute_message_hash([context.msg_sender(), selector, from.address, to.address, amount, nonce]);
+            assert(AccountContract::at(from.address).is_valid(Option::none(), Option::some(context), message_field) == 0xe86ab4ff, "invalid call");
```

Comment:
FYI I had added a helper for this here: https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/noir-libs/noir-aztec/src/auth.nr. Maybe we can include it in your account contract interface and remove it as a separate util?

Reply:
Ah ok, seems nice. Will update to use that one instead.

Reply:
Reviewed, and think I prefer the combined `is_valid` that handles both the private and public flows with same interface more 🤔

Reply:
Agree, but I think there's value in a helper that removes the need for hardcoding the `0xe86ab4ff` selector in your app logic. Maybe add that check to the account interface?